Decimal reagents

PrPleGoo opened the pull request:
https://github.com/space-wizards/space-station-14/issues/377

Changes all ints in and around reagents to decimals.
Adds a special rounder for reagents.
Change some constructor calls for Solution to inject rounder dependency.
Renamed some fields to not break convention.

Can't merge because it relies on changes in RobustToolbox: https://github.com/space-wizards/RobustToolbox/pull/1021

PJB3005 commented:
Why are you using floating point? You can just use fixed point to like 2 or 3 decimal places to get more than enough precision anybody could ever need, without the jank of getting 3.5e-10 oxygen in your beaker. `decimal` doesn't exactly help with this..

PrPleGoo commented:
please restart travis for this PR

PJB3005 commented on Content.Shared/Chemistry/ReagentUnit.cs:
```diff
@@ -0,0 +1,184 @@
+            return New(FloatFromString(value));
+        }
+
+        private static float FloatFromString(string value)
```
You're gonna need a way to pass through the `CultureInfo` or such too here, because right now this uses current-culture parsing meaning that `Deserialize` doesn't work correctly when your culture is, say, french.

PrPleGoo replied:
1. Added
2. Not sure what you mean by this but I added some unit tests for some nasty fractional cases and they all seem to be handled well.

PrPleGoo replied:
I'm just passing invariant culture to all the string things./

PJB3005 commented on Content.Shared/Chemistry/ReagentUnit.cs:
```diff
@@ -0,0 +1,184 @@
+namespace Content.Shared.Chemistry
+{
+    [Serializable]
+    public struct ReagentUnit : ISelfSerialize
```
You may want to implement some common interfaces like `IEquatable<ReagentUnit>` and `IComparable<ReagentUnit>` here.

PrPleGoo replied:
Done

PJB3005 commented on Content.Shared/Chemistry/ReagentUnit.cs:
```diff
@@ -0,0 +1,184 @@
+
+        public static ReagentUnit MaxValue => new ReagentUnit(int.MaxValue);
+
+        private decimal ShiftDown()
```
Not sure whether you need the full precision of `decimal` when working with this. IIRC, `double` can accurately represent 32 bit-integers completely so that should be good enough. It'd help with performance quite a bit.

PrPleGoo replied:
done

PJB3005 commented on Content.Shared/Chemistry/ReagentUnit.cs:
```diff
@@ -0,0 +1,184 @@
+            return (int) ShiftDown();
+        }
+
+        public static ReagentUnit Min(params ReagentUnit[] reagentUnits)
```
1. you might want an overload that takes in only two plain parameters so that you don't always need to allocate when using this method.
2. Using LINQ for this also adds quite a bit of allocations and slowdown compared to just a plain loop.
2.1 <sup>LINQ has a `.Min` method</sup>

PrPleGoo replied:
1. done
2.1 used that instead

PJB3005 commented on Content.Shared/Chemistry/ReagentUnit.cs:
```diff
@@ -0,0 +1,184 @@
+
+        public string Serialize()
+        {
+            return ToString();
```
This should also use invariant culture.

PrPleGoo commented:
> Looks decent, although I'd like to see some more accurate-fixed-point APIs:
> 
> 1. A multiply with an int, since you can do that without having to shift down and back up.
> 2. A division with remainder, useful to split a beaker in half (the .01 has to go somewhere and mustn't be forgotten about).

Added
Not sure what you mean by this but I added some unit tests for some nasty fractional cases and they all seem to be handled well.
